tornado/test/testing_test.py

```diff
@@ -228,5 +248,31 @@ def test_with_kwargs(self, **kwargs):
         test_with_kwargs(self, test='test')
         self.finished = True
 
+    @skipBefore35
+    def test_native_coroutine(self):
+        namespace = exec_test(globals(), locals(), """
+        @gen_test
+        async def test(self):
+            self.finished = True
+        """)
+
+        namespace['test'](self)
+
+    @skipBefore35
+    def test_native_coroutine_timeout(self):
+        # Set a short timeout and exceed it.
+        namespace = exec_test(globals(), locals(), """
+        @gen_test(timeout=0.1)
+        async def test(self):
+            await gen.sleep(1)
+        """)
+
+        try:
+            namespace['test'](self)
+            self.fail("did not get expected exception")
+        except ioloop.TimeoutError:
+            self.finished = True
```

Comment:
The traceback manipulation works for native test coroutines, same as for generator-based test coroutines. In Python 3.5 this coroutine's failure traceback includes "await gen.sleep(1)".

It's hard to test, however, because an exec'ed string doesn't give a good traceback under any circumstances. We could move this test to its own file and import it in Python 3.5+, if we're committed to testing timeout tracebacks from native test coroutines.